gentypes.go

```diff
@@ -0,0 +1,211 @@
+package gosysl
+
+import (
+	"bytes"
+	"fmt"
+	"go/format"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/anz-bank/gosysl/pb"
+)
+
+// GenPrimitiveType return Golang type string for Sysl primitive data type
+func GenPrimitiveType(tp *pb.Type_Primitive) (string, error) {
+	switch *tp {
+	case pb.Type_BOOL:
+		return "bool", nil
+	case pb.Type_ANY:
+		return "interface{}", nil
+	case pb.Type_INT:
+		return "int", nil
+	case pb.Type_STRING:
+		return "string", nil
+	case pb.Type_EMPTY:
+		return "nil", nil
+	case pb.Type_FLOAT, pb.Type_DECIMAL:
+		return "float64", nil
+	case pb.Type_BYTES:
+		return "[]byte", nil
+	case pb.Type_DATE, pb.Type_DATETIME:
+		return "time.Time", nil
+	}
+	// Type_XML, Type_UUID
+	return "", fmt.Errorf("unsupported type primitive %s", tp.String())
+}
+
+// GenSimpleType returns Golang type string for non-composite types (no lists and sets)
+func GenSimpleType(t *pb.Type) (string, error) {
+	if pType := t.GetPrimitive(); pType != pb.Type_NO_Primitive {
+		return GenPrimitiveType(&pType)
+	}
+	if t.GetTypeRef() != nil {
+		path := t.GetTypeRef().GetRef().GetPath()
+		if len(path) != 1 {
+			return "", fmt.Errorf("cannot handle type reference with more than one path")
+		}
+		str := path[0]
+		if !strings.HasPrefix(str, "map of") {
+			return str, nil
+		}
+		str = strings.TrimPrefix(str, "map of")
+		str = strings.Trim(str, " ")
+		m := strings.Split(str, ":")
+		if len(m) != 2 {
+			shouldStr := `should be map of KeyType:ValueType`
+			return "", fmt.Errorf("bad map definition '%s' (%s)", str, shouldStr)
+		}
+		return fmt.Sprintf("map[%s]%s", m[0], m[1]), nil
+	}
+	return "", fmt.Errorf("type %v is neither primitive nor reference", t)
+
+}
+
+// GetLine returns the line for a given Sysl type from its SourceContext
+func GetLine(t *pb.Type) (int32, error) {
+	if t.GetPrimitive() != pb.Type_NO_Primitive || t.GetTypeRef() != nil {
+		return t.GetSourceContext().Start.Line, nil
+	} else if t.GetList() != nil {
```

Comment:
No `else` required. Ditto below. I believe golint would normally complain, except that (iirc) it gives up on multistage if-else chains. I will probably start complaining in future, though.